src/arithmetic/const_int_bound.cc

```diff
@@ -387,6 +428,36 @@ class ConstIntBoundAnalyzer::Impl :
     }
     return ret;
   }
+
+  /*!
+   * \brief Detect additional constant bound from cond, if any
+   * \param cond The constraint condition.
+   * \return List of detected bounds.
+   */
+  static std::vector<BoundInfo> DetectBoundInfo(const Expr& cond) {
+    PVar<Expr> x, y;
+    PVar<Integer> c;
+    // NOTE: canonical form always use <= or <
+    if ((c <= x).Match(cond)) {
+      return {BoundInfo(x.Eval(), MakeBound(c.Eval()->value, kPosInf))};
+    }
+    if ((c < x).Match(cond)) {
+      return {BoundInfo(x.Eval(), MakeBound(c.Eval()->value + 1, kPosInf))};
+    }
+    if ((x <= c).Match(cond)) {
+      return {BoundInfo(x.Eval(), MakeBound(kNegInf, c.Eval()->value))};
+    }
+    if ((x < c).Match(cond)) {
+      return {BoundInfo(x.Eval(), MakeBound(kNegInf, c.Eval()->value - 1))};
+    }
+    if ((x && y).Match(cond)) {
+      auto ret1 = DetectBoundInfo(x.Eval());
+      auto ret2 = DetectBoundInfo(y.Eval());
+      ret1.insert(ret1.end(), ret2.begin(), ret2.end());
+      return ret1;
+    }
```

Comment:
Do we also need the `x == c` (and `c == x`) case?

Reply:
I am not very sure if this is a case that we will need as frequently, so I decided to leave it out after some thought. We can revisit later. This might be better suited for a condition in a simplifier. Most conditions like ```x % 3== 1``` is detected by modular analysis.